### PR TITLE
fix(memory): handle BOM and non-UTF-8 bytes in memory files

### LINKED ISSUE #10878: [Bug]: memory_tool _read_file does not strip BOM — invisible \ufeff enters system prompt
## Bug Description

`MemoryStore._read_file()` in `tools/memory_tool.py` (line 419) reads with `encoding="utf-8"` which does **not** auto-strip a BOM (`\ufeff`). If MEMORY.md starts with a UTF-8 BOM (common when edited on Windows with Notepad), the `\ufeff` character becomes the first character of the first entry.

The BOM is listed in `_INVISIBLE_CHARS` (line 72), so it would be blocked if written through the built-in `add` path. But `_read_file` does **not** call `_scan_memory_content`, so the BOM survives into the snapshot and is injected into the system prompt.

## Steps to Reproduce

```bash
python3 -c "
from pathlib import Path
p = Path.home() / '.hermes' / 'memories' / 'MEMORY.md'
p.write_text('\ufeffUser prefers dark mode', encoding='utf-8')
"
# Start hermes — BOM is in the system prompt
```

## Expected Behavior

`_read_file` should strip the BOM on load, or `load_from_disk` should strip `\ufeff` from the raw content before splitting.

A simple fix:
```python
raw = path.read_text(encoding="utf-8-sig")  # auto-strips BOM
```

## Additional Context

This is particularly relevant for Windows users where text editors commonly add BOM to UTF-8 files, and for external MCP writers that may inadvertently produce BOM-prefixed output.

Related: #8450, #8467 (UTF-8 handling in plugin.yaml, same class of issue)

### LINKED ISSUE #10879: [Bug]: Non-UTF-8 MEMORY.md raises uncaught UnicodeDecodeError — agent starts with empty memory
## Bug Description

`MemoryStore._read_file()` in `tools/memory_tool.py` (line 419) calls `path.read_text(encoding="utf-8")` without an `errors=` parameter. If MEMORY.md contains non-UTF-8 bytes, this raises `UnicodeDecodeError`.

The exception propagates through `load_from_disk` and is caught by a bare `except Exception: pass` in `run_agent.py` (lines 1229–1230). The result: `_memory_store` remains with empty entries, and the **agent silently starts the session as if it has no memory**.

No warning is logged. The user has no indication that their memory was silently dropped.

## Steps to Reproduce

```bash
# Write a file with Latin-1 encoding (non-UTF-8)
python3 -c "
from pathlib import Path
p = Path.home() / '.hermes' / 'memories' / 'MEMORY.md'
p.write_bytes(b'User prefers caf\xe9 mode')
"
# Start hermes — agent has empty memory, no error shown
hermes chat -q "What do you remember?"
```

## Expected Behavior

Either:
1. Read with `errors="replace"` or `errors="surrogateescape"` so content loads with replacement characters
2. Or log a clear warning: `"WARNING: MEMORY.md contains non-UTF-8 bytes, memory may be incomplete"`

The current silent empty-memory fallback is the worst outcome — the user loses all memory with no diagnostic.

## Additional Context

This can happen when:
- A user edits MEMORY.md in an editor with non-UTF-8 locale (common on Windows)
- An external tool writes to MEMORY.md with the wrong encoding
- File corruption introduces invalid bytes

Related: #8898, #9145 (UnicodeDecodeError in migration scripts — same class of bug, different location)

### PR DESCRIPTION
**Pull request by easyvibecoding**
## What does this PR do?

Change `_read_file` encoding from `utf-8` to `utf-8-sig` (auto-strips BOM) and add `errors="replace"` to handle non-UTF-8 bytes gracefully.

## Related Issue

Fixes #10878, fixes #10879

## Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)

## Changes Made

- `tools/memory_tool.py` line 419: `path.read_text(encoding="utf-8-sig", errors="replace")`

**Before:**
- BOM (`\ufeff`) at file start survived into the system prompt, potentially confusing the model
- Non-UTF-8 bytes raised `UnicodeDecodeError`, which was caught by a bare `except Exception: pass` in `run_agent.py:1229`, causing the agent to start with **completely empty memory and no diagnostic**

**After:**
- BOM is auto-stripped by the `utf-8-sig` codec
- Non-UTF-8 bytes are replaced with U+FFFD instead of crashing

## How to Test

```bash
# Test BOM handling
python3 -c "
from pathlib import Path
p = Path.home() / '.hermes' / 'memories' / 'MEMORY.md'
p.write_text('\ufeffUser prefers dark mode', encoding='utf-8')
"
# Start hermes — entry should load without leading BOM character

# Test non-UTF-8 handling
python3 -c "
from pathlib import Path
p = Path.home() / '.hermes' / 'memories' / 'MEMORY.md'
p.write_bytes(b'User prefers caf\xe9 mode')
"
# Start hermes — entry should load with replacement char, not empty memory
```

## Checklist

- [x] I've read the [Contributing Guide](https://github.com/NousResearch/hermes-agent/blob/main/CONTRIBUTING.md)
- [x] My commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)
- [x] My PR contains **only** changes related to this fix (1 file, 5 lines changed)
- [x] I've considered cross-platform impact — BOM is most common on Windows, this fix is platform-neutral
- [x] I've tested on my platform: macOS 15, Python 3.12.8

### CODE REVIEW

**Comment by mxnstrexgl** [comment]
## 🔒 Security & Code Quality Review

**Status:** ✅ No security issues found. Minor code quality observation.

---

### Security Analysis

| Check | Result |
|-------|--------|
| Hardcoded secrets | ✅ None found |
| Path traversal | ✅ Mitigated - paths constructed via , validated with  |
| Injection vulnerabilities | ✅ None - no SQL, shell, or other injection vectors |
| Unsafe deserialization | ✅ N/A - text file reading only |
| Error handling | ✅ Properly catches  specifically |

### Code Quality Findings

**1. Missing Test Coverage for New Behavior** ⚠️
The PR introduces BOM handling and non-UTF-8 byte replacement but doesn't include automated tests. While the PR description provides excellent manual test instructions, consider adding a test to :



**2. Data Integrity Consideration** 📋
Using  will silently replace invalid UTF-8 bytes with . This is a reasonable trade-off (partial data > no data), but consider logging when replacement occurs:



### Positive Observations

- ✅ The  codec correctly handles Windows BOM files
- ✅ Comments clearly explain the rationale for the change
- ✅ All 33 existing tests pass with the change
- ✅ Exception handling is specific and appropriate (catches  not bare )
- ✅ The fix addresses the root cause: UnicodeDecodeError causing empty memory

---

**Overall:** This is a solid bugfix with good security posture. The only recommendation is to add automated tests for the new edge case handling.

LGTM with minor suggestions for future improvement. 🚀

File: tools/memory_tool.py [hunk 1]
```diff
@@ -416,7 +416,11 @@ def _read_file(path: Path) -> List[str]:
         if not path.exists():
             return []
         try:
-            raw = path.read_text(encoding="utf-8")
+            # Use utf-8-sig to auto-strip BOM if present (common on Windows).
+            # Fall back to replacement characters for non-UTF-8 bytes rather
+            # than raising UnicodeDecodeError (which was previously swallowed
+            # by a bare except, silently emptying all memory).
+            raw = path.read_text(encoding="utf-8-sig", errors="replace")
```
Comment (teknium1) [inline tools/memory_tool.py:706]: Please apply the same decoding policy in `_detect_external_drift()`. Current main calls that helper before `_read_file()` for replace/remove/batch mutations (`tools/memory_tool.py:303,403,464,524`), and its strict UTF-8 read can still raise on the files this line makes loadable.

**Review by teknium1 — COMMENTED** [review]
Thank you for the focused regression tests. The startup-load premise remains valid on current main: `tools/memory_tool.py:692` strictly decodes UTF-8, and `agent/agent_init.py:1345-1347` silently absorbs failures during memory bootstrap.

### Problems
- The fix does not cover mutations. `_reload_target()` calls `_detect_external_drift()` first (`tools/memory_tool.py:303`), while that helper still reads with strict UTF-8 at `tools/memory_tool.py:732`. `replace`, `remove`, and `apply_batch` all use this path (`:403`, `:464`, `:524`), so invalid bytes can still escape before the changed reader runs.
- A BOM would also cause `_detect_external_drift()`'s round-trip comparison (`tools/memory_tool.py:738-744`) to report drift during these mutations.

### Suggested changes
- Normalize `_detect_external_drift()` with the same `utf-8-sig` plus `errors="replace"` policy.
- Extend the new load tests with a replace/remove or batch mutation against BOM-prefixed and invalid-byte files.

This is an automated hermes-sweeper review.

File: tools/memory_tool.py (hunk 1 above)
Comment (easyvibecoding) [reply tools/memory_tool.py:706]: Applied in fcdfaf58e: `_detect_external_drift()` now reads with `utf-8-sig` + `errors="replace"`, same policy as `_read_file()`. Branch rebased onto current main. Added mutation coverage: replace on a BOM-prefixed file, remove on an invalid-byte file, apply_batch on a BOM+invalid-byte file (tests/tools/test_memory_tool.py, TestMemoryStoreEncoding). 90/90 tests pass locally.

One correction to the review note: on current main a BOM does not trip the round-trip comparison. `str.strip()` does not remove U+FEFF (`"\ufeff".isspace()` is False), so the BOM survives identically in both `raw.strip()` and the re-serialized string — they still compare equal. The actual failure mode was the strict decode raising `UnicodeDecodeError` (a `ValueError`, not caught by the `(OSError, IOError)` handler) on the replace/remove/batch paths before the resilient reader ran. Normalizing both readers to one policy fixes that and keeps the two views of the file consistent.

**Comment by teknium1** [comment]
Salvaged into #81967 — the memory BOM handling landed with Co-authored-by credit (the strict-decode contract on main had evolved past a clean cherry-pick). Fixes the \ufeff-glued-to-first-entry bug from #10878. Thanks!
